# Analyzing _Ebay_ Car Sales Data

This project will focus on a used car dataset scraped from _eBay Kleinanzeigen_, a "classifieds" section of the German _eBay_. Originally scraped, cleaned, and uploaded to Kaggle by the user **orgesleka** (deleted). Original dataset can be found on _data.world_  <a href="https://data.world/data-society/used-cars-data">here</a>.

The data worked here is a "dirtied" sample of the original meant to more accurately represent a scraped dataset.

### Goal:

To clean and analyze the dataset, noting any observations.

In [1]:
import pandas as pd
import numpy as np

from datetime import datetime as dt

from IPython.display import display

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
display(autos.info())
display(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

None

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

#### Note:
Field values are in German. Column names in camel case. 

In [4]:
# Examine before and after of some passed operation

def show_change(op, **kwargs):
    before = autos
    
    op(**kwargs)
    
    print("BEFORE:")
    display(before.head())
    print("AFTER:")
    display(autos.head())
    
    
    print("\n\nBEFORE:")
    display(before.describe(include="all"))
    print("AFTER:")
    display(autos.describe(include="all"))

In [5]:
# Let's clean the data

col_map = {"dateCrawled":"date_crawled",
    "offerType":"offer_type",
    "vehicleType":"type",
    "yearOfRegistration":"registration_year",
    "powerPS":"power_ps",
    "monthOfRegistration":"registration_month",
    "fuelType":"fuel",
    "notRepairedDamage":"unrepaired_damage",
    "dateCreated":"ad_created",
    "postalCode":"postal_code",
    "lastSeen":"last_update"}

show_change(autos.rename, mapper=col_map, axis=1, inplace=True)

BEFORE:


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest        type  registration_year  \
0  privat    Angebot  $5,000  control         bus               2004   
1  privat    Angebot  $8,500  control   limousine               1997   
2  privat    Angebot  $8,990     test   limousine               2009   
3  privat    Angebot  $4,350  control  kleinwagen               2007   
4  privat    Angebot  $1,350     test       kombi               2003   

     gearbox  power_ps   model   odometer  registration_month    fuel  \
0    manuell       158  andere  150,000km                   3     lpg   
1  automatik       286     7er  150,000km                   6  benzin   
2    manuell       102    golf   70,000km                   7  benzin   
3  automatik        71  fortwo   70,000km                   6  benzin   
4    manuell         0   focus  150,000km                   7  benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code          last_update  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

AFTER:


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest        type  registration_year  \
0  privat    Angebot  $5,000  control         bus               2004   
1  privat    Angebot  $8,500  control   limousine               1997   
2  privat    Angebot  $8,990     test   limousine               2009   
3  privat    Angebot  $4,350  control  kleinwagen               2007   
4  privat    Angebot  $1,350     test       kombi               2003   

     gearbox  power_ps   model   odometer  registration_month    fuel  \
0    manuell       158  andere  150,000km                   3     lpg   
1  automatik       286     7er  150,000km                   6  benzin   
2    manuell       102    golf   70,000km                   7  benzin   
3  automatik        71  fortwo   70,000km                   6  benzin   
4    manuell         0   focus  150,000km                   7  benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postal_code          last_update  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50



BEFORE:


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

             type  registration_year  gearbox      power_ps  model   odometer  \
count       44905       50000.000000    47320  50000.000000  47242      50000   
unique          8                NaN        2           NaN    245         13   
top     limousine                NaN  manuell           NaN   golf  150,000km   
freq        12859                NaN    36993           NaN   4024      32424   
mean          NaN        2005.073280      NaN    116.355920    NaN        NaN   
std           NaN         105.712813      NaN    209.216627    NaN        NaN   
min           NaN        1000.000000      NaN      0.000000    NaN        NaN   
25%           NaN        1999.000000      NaN     70.000000    NaN        NaN   
50%           NaN        2003.000000      NaN    105.000000    NaN        NaN   
75%           NaN        2008.000000      NaN    150.000000    NaN        NaN   
max           NaN        9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month    fuel       brand unrepaired_damage  \
count         50000.000000   45518       50000             40171   
unique                 NaN       7          40                 2   
top                    NaN  benzin  volkswagen              nein   
freq                   NaN   30107       10687             35232   
mean              5.723360     NaN         NaN               NaN   
std               3.711984     NaN         NaN               NaN   
min               0.000000     NaN         NaN               NaN   
25%               3.000000     NaN         NaN               NaN   
50%               6.000000     NaN         NaN               NaN   
75%               9.000000     NaN         NaN               NaN   
max              12.000000     NaN         NaN               NaN   

                 ad_created  nrOfPictures   postal_code          last_update  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

AFTER:


date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

             type  registration_year  gearbox      power_ps  model   odometer  \
count       44905       50000.000000    47320  50000.000000  47242      50000   
unique          8                NaN        2           NaN    245         13   
top     limousine                NaN  manuell           NaN   golf  150,000km   
freq        12859                NaN    36993           NaN   4024      32424   
mean          NaN        2005.073280      NaN    116.355920    NaN        NaN   
std           NaN         105.712813      NaN    209.216627    NaN        NaN   
min           NaN        1000.000000      NaN      0.000000    NaN        NaN   
25%           NaN        1999.000000      NaN     70.000000    NaN        NaN   
50%           NaN        2003.000000      NaN    105.000000    NaN        NaN   
75%           NaN        2008.000000      NaN    150.000000    NaN        NaN   
max           NaN        9999.000000      NaN  17700.000000    NaN        NaN   

        registration_month    fuel       brand unrepaired_damage  \
count         50000.000000   45518       50000             40171   
unique                 NaN       7          40                 2   
top                    NaN  benzin  volkswagen              nein   
freq                   NaN   30107       10687             35232   
mean              5.723360     NaN         NaN               NaN   
std               3.711984     NaN         NaN               NaN   
min               0.000000     NaN         NaN               NaN   
25%               3.000000     NaN         NaN               NaN   
50%               6.000000     NaN         NaN               NaN   
75%               9.000000     NaN         NaN               NaN   
max              12.000000     NaN         NaN               NaN   

                 ad_created  nrOfPictures   postal_code          last_update  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [11]:
# More detailed output on specific columns

def col_detail(bound=0.0, *col):
    
    results = {}
    
    for c in col:
        series = None
        
        if (bound < 0.0) or (bound > 1.0):
            raise ValueError("Bad bound; must be a fraction of 1")
                
        limit = int(autos[c].shape[0]*bound)
            
        if 0.0 < bound < 0.5:
            series = autos[c].sort_values(ascending=True)[:limit]
        elif 1.0 > bound >= 0.5:
            series = autos[c].sort_values(ascending=True)[limit:]
        else:
            series = autos[c]
            
        results[c] = series
        
        print("\n\nColumn ({}):"
              .format(c.upper()))
        print("\tUnique values: \n{}"
              .format(series.unique().shape))
        print("\tDescription: \n{}"
              .format(series.describe()))
        print("\tValue counts: \n"
              "\n\t\tGreatest: \n{}"
              "\n\t\tLeast: \n{}"
              .format(
                  series.value_counts().sort_index(ascending=False).head(),
                  series.value_counts().sort_index(ascending=True).head()))
        
    return results

In [12]:
col_detail(0.0, *[c for c in autos.columns])



Column (DATE_CRAWLED):
	Unique values: 
(48213,)
	Description: 
count                   50000
unique                  48213
top       2016-03-12 16:06:22
freq                        3
Name: date_crawled, dtype: object
	Value counts: 

		Greatest: 
2016-04-07 14:36:56    1
2016-04-07 14:36:55    1
2016-04-07 14:36:44    1
2016-04-07 14:30:26    1
2016-04-07 14:30:09    1
Name: date_crawled, dtype: int64
		Least: 
2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:04    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
Name: date_crawled, dtype: int64


Column (NAME):
	Unique values: 
(38754,)
	Description: 
count           50000
unique          38754
top       Ford_Fiesta
freq               78
Name: name, dtype: object
	Value counts: 

		Greatest: 
Älteres_Auto_zu_Verkaufen_aber_Laeuft.                          1
»_TWINGO_«                                                      1
°_°_MERCEDES_C_KLASSE_180_MIT_TÜV_°_°                           1
°_Mercedes_Benz_190_E_2.

{'abtest': 0        control
 1        control
 2           test
 3        control
 4           test
 5           test
 6           test
 7        control
 8           test
 9        control
 10          test
 11       control
 12       control
 13       control
 14          test
 15          test
 16          test
 17          test
 18       control
 19       control
 20          test
 21          test
 22       control
 23       control
 24       control
 25       control
 26       control
 27       control
 28       control
 29          test
           ...   
 49970    control
 49971       test
 49972    control
 49973    control
 49974    control
 49975    control
 49976       test
 49977    control
 49978    control
 49979       test
 49980    control
 49981    control
 49982    control
 49983       test
 49984       test
 49985    control
 49986    control
 49987    control
 49988    control
 49989       test
 49990       test
 49991    control
 49992    control
 49993    control


#### Notes:
* Above translation map must be applied to change German to English
* ***nrOfPictures***: can be deleted, all entries are 0 
* ***name***: values changed to proper snake case, remove special characters
* ***price***: change to ***price_dollar***, remove '$', change dtype to _int_
* ***abtest***: is an unknown, keep
* ***odometer***: change to ***odometer_km***, remove 'km' from values, change dtype to _int_
* ***registration_month, registration_year***: change month of 0 -> 1, combine into one column (***date_registration***)
* ***date_crawled, ad_created***: change to ***date_ad_created***, change to _datetime_
* ***registration_year, power_ps***: bad values, examine further

In [13]:
autos.drop(columns="nrOfPictures", inplace=True)

In [14]:
rename = {"price": "price_dollar",
         "odometer": "odometer_km",
         "ad_created": "date_ad_created"}

autos["name"] = autos["name"].str.lower().str.replace(r"([^a-zA-Z0-9_./]+|[./_]{2,})", "")
autos["price"] = autos["price"].str.replace(r"[$,]+", "")
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].str.replace(r"[km,]", "")
autos["odometer"] = autos["odometer"].astype(int)
autos.rename(columns=rename, inplace=True)

In [15]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   peugeot_807_160_navtech_on_board   
1  2016-04-04 13:38:56         bmw_740i_4_4_liter_hamann_umbau_mega_optik   
2  2016-03-26 18:57:24                         volkswagen_golf_1.6_united   
3  2016-03-12 16:58:10  smart_smart_fortwo_coupe_softouch/f1/klima/pan...   
4  2016-04-01 14:38:50  ford_focus_1_6_benzin_tv_neu_ist_sehr_gepflegt...   

   seller offer_type  price_dollar   abtest        type  registration_year  \
0  privat    Angebot          5000  control         bus               2004   
1  privat    Angebot          8500  control   limousine               1997   
2  privat    Angebot          8990     test   limousine               2009   
3  privat    Angebot          4350  control  kleinwagen               2007   
4  privat    Angebot          1350     test       kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month    fuel  \
0    manuell       158  andere       150000                   3     lpg   
1  automatik       286     7er       150000                   6  benzin   
2    manuell       102    golf        70000                   7  benzin   
3  automatik        71  fortwo        70000                   6  benzin   
4    manuell         0   focus       150000                   7  benzin   

        brand unrepaired_damage      date_ad_created  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00        79588   
1         bmw              nein  2016-04-04 00:00:00        71034   
2  volkswagen              nein  2016-03-26 00:00:00        35394   
3       smart              nein  2016-03-12 00:00:00        33729   
4        ford              nein  2016-04-01 00:00:00        39218   

           last_update  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [16]:
year = autos["registration_year"].astype(str)
month = autos["registration_month"].astype(str).replace("0", np.nan)

autos["date_registration"] = year + "-" + month

autos.drop(columns=["registration_year", "registration_month"], inplace=True)

In [17]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   peugeot_807_160_navtech_on_board   
1  2016-04-04 13:38:56         bmw_740i_4_4_liter_hamann_umbau_mega_optik   
2  2016-03-26 18:57:24                         volkswagen_golf_1.6_united   
3  2016-03-12 16:58:10  smart_smart_fortwo_coupe_softouch/f1/klima/pan...   
4  2016-04-01 14:38:50  ford_focus_1_6_benzin_tv_neu_ist_sehr_gepflegt...   

   seller offer_type  price_dollar   abtest        type    gearbox  power_ps  \
0  privat    Angebot          5000  control         bus    manuell       158   
1  privat    Angebot          8500  control   limousine  automatik       286   
2  privat    Angebot          8990     test   limousine    manuell       102   
3  privat    Angebot          4350  control  kleinwagen  automatik        71   
4  privat    Angebot          1350     test       kombi    manuell         0   

    model  odometer_km    fuel       brand unrepaired_damage  \
0  andere       150000     lpg     peugeot              nein   
1     7er       150000  benzin         bmw              nein   
2    golf        70000  benzin  volkswagen              nein   
3  fortwo        70000  benzin       smart              nein   
4   focus       150000  benzin        ford              nein   

       date_ad_created  postal_code          last_update date_registration  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54            2004-3  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08            1997-6  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37            2009-7  
3  2016-03-12 00:00:00        33729  2016-03-15 03:16:28            2007-6  
4  2016-04-01 00:00:00        39218  2016-04-01 14:38:50            2003-7

In [18]:
format_date = "%Y-%m-%d %H:%M:%S"
format_date_reg = "%Y-%m"

date_crawled = pd.to_datetime(autos["date_crawled"], format=format_date)
date_ad = pd.to_datetime(autos["date_ad_created"], format=format_date)
date_reg = pd.to_datetime(autos["date_registration"], errors="coerce", format=format_date_reg)

display(date_crawled.head())
display(date_ad.head())
display(date_reg.head())

0   2016-03-26 17:47:46
1   2016-04-04 13:38:56
2   2016-03-26 18:57:24
3   2016-03-12 16:58:10
4   2016-04-01 14:38:50
Name: date_crawled, dtype: datetime64[ns]

0   2016-03-26
1   2016-04-04
2   2016-03-26
3   2016-03-12
4   2016-04-01
Name: date_ad_created, dtype: datetime64[ns]

0   2004-03-01
1   1997-06-01
2   2009-07-01
3   2007-06-01
4   2003-07-01
Name: date_registration, dtype: datetime64[ns]

In [19]:
autos["date_crawled"] = date_crawled
autos["date_ad_created"] = date_ad
autos["date_registration"] = date_reg

autos.describe(include="all")

date_crawled        name  seller offer_type  price_dollar  \
count                 50000       50000   50000      50000  5.000000e+04   
unique                48213       37775       2          2           NaN   
top     2016-03-12 16:06:22  opel_corsa  privat    Angebot           NaN   
freq                      3          97   49999      49999           NaN   
first   2016-03-05 14:06:30         NaN     NaN        NaN           NaN   
last    2016-04-07 14:36:56         NaN     NaN        NaN           NaN   
mean                    NaN         NaN     NaN        NaN  9.840044e+03   
std                     NaN         NaN     NaN        NaN  4.811044e+05   
min                     NaN         NaN     NaN        NaN  0.000000e+00   
25%                     NaN         NaN     NaN        NaN  1.100000e+03   
50%                     NaN         NaN     NaN        NaN  2.950000e+03   
75%                     NaN         NaN     NaN        NaN  7.200000e+03   
max                     NaN         NaN     NaN        NaN  1.000000e+08   

       abtest       type  gearbox      power_ps  model    odometer_km    fuel  \
count   50000      44905    47320  50000.000000  47242   50000.000000   45518   
unique      2          8        2           NaN    245            NaN       7   
top      test  limousine  manuell           NaN   golf            NaN  benzin   
freq    25756      12859    36993           NaN   4024            NaN   30107   
first     NaN        NaN      NaN           NaN    NaN            NaN     NaN   
last      NaN        NaN      NaN           NaN    NaN            NaN     NaN   
mean      NaN        NaN      NaN    116.355920    NaN  125732.700000     NaN   
std       NaN        NaN      NaN    209.216627    NaN   40042.211706     NaN   
min       NaN        NaN      NaN      0.000000    NaN    5000.000000     NaN   
25%       NaN        NaN      NaN     70.000000    NaN  125000.000000     NaN   
50%       NaN        NaN      NaN    105.000000    NaN  150000.000000     NaN   
75%       NaN        NaN      NaN    150.000000    NaN  150000.000000     NaN   
max       NaN        NaN      NaN  17700.000000    NaN  150000.000000     NaN   

             brand unrepaired_damage      date_ad_created   postal_code  \
count        50000             40171                50000  50000.000000   
unique          40                 2                   76           NaN   
top     volkswagen              nein  2016-04-03 00:00:00           NaN   
freq         10687             35232                 1946           NaN   
first          NaN               NaN  2015-06-11 00:00:00           NaN   
last           NaN               NaN  2016-04-07 00:00:00           NaN   
mean           NaN               NaN                  NaN  50813.627300   
std            NaN               NaN                  NaN  25779.747957   
min            NaN               NaN                  NaN   1067.000000   
25%            NaN               NaN                  NaN  30451.000000   
50%            NaN               NaN                  NaN  49577.000000   
75%            NaN               NaN                  NaN  71540.000000   
max            NaN               NaN                  NaN  99998.000000   

                last_update    date_registration  
count                 50000                44918  
unique                39481                  666  
top     2016-04-07 06:17:27  2003-03-01 00:00:00  
freq                      8                  376  
first                   NaN  1800-02-01 00:00:00  
last                    NaN  2019-05-01 00:00:00  
mean                    NaN                  NaN  
std                     NaN                  NaN  
min                     NaN                  NaN  
25%                     NaN                  NaN  
50%                     NaN                  NaN  
75%                     NaN                  NaN  
max                     NaN                  NaN

In [20]:
# Dict extension class that returns any missing values instead of NaN
class Translations(dict):
    def __missing__(self, key):
        return key

# German:English translation map
translations = Translations([("privat", "private"),
               ("gewerblich", "commercial"),
               ("Angebot", "offer"),
               ("Gesuch", "request"),
               ("kleinwagen", "small"),
               ("kombi", "combi"),
               ("cabrio", "convertible"),
               ("andere", "other"),
               ("manuell", "manual"),
               ("automatik", "automatic"),
               ("benzin", "petrol"),
               ("elektro", "electric"),
               ("nein", "no"),
               ("ja", "yes")])

In [21]:
for col in autos.columns:
    autos[col] = autos[col].map(translations, na_action="ignore")

In [23]:
autos.head()

col_detail(0.0, *[c for c in autos.columns])



Column (DATE_CRAWLED):
	Unique values: 
(48213,)
	Description: 
count                   50000
unique                  48213
top       2016-03-12 16:06:22
freq                        3
first     2016-03-05 14:06:30
last      2016-04-07 14:36:56
Name: date_crawled, dtype: object
	Value counts: 

		Greatest: 
2016-04-07 14:36:56    1
2016-04-07 14:36:55    1
2016-04-07 14:36:44    1
2016-04-07 14:30:26    1
2016-04-07 14:30:09    1
Name: date_crawled, dtype: int64
		Least: 
2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:04    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
Name: date_crawled, dtype: int64


Column (NAME):
	Unique values: 
(37775,)
	Description: 
count          50000
unique         37775
top       opel_corsa
freq              97
Name: name, dtype: object
	Value counts: 

		Greatest: 
zwischenloesung                                   1
zweimassenschwungradraederzws_crafter_sprinter    1
zwei_fahrzeuges_gegen_eine                        1
zwar_aelte

{'abtest': 0        control
 1        control
 2           test
 3        control
 4           test
 5           test
 6           test
 7        control
 8           test
 9        control
 10          test
 11       control
 12       control
 13       control
 14          test
 15          test
 16          test
 17          test
 18       control
 19       control
 20          test
 21          test
 22       control
 23       control
 24       control
 25       control
 26       control
 27       control
 28       control
 29          test
           ...   
 49970    control
 49971       test
 49972    control
 49973    control
 49974    control
 49975    control
 49976       test
 49977    control
 49978    control
 49979       test
 49980    control
 49981    control
 49982    control
 49983       test
 49984       test
 49985    control
 49986    control
 49987    control
 49988    control
 49989       test
 49990       test
 49991    control
 49992    control
 49993    control


#### Note:
Translations and formatting errors addressed. Further analysis showing that columns ***seller, offer_type*** are predominantly single-valued (49,999:1); need to be dropped. Also need to deal with outliers.

In [24]:
autos.drop(columns=["seller", "offer_type"], inplace=True)

In [26]:
# First deal with columns odometer_km and price_dollar

col_detail(0.0, "odometer_km", "price_dollar")



Column (ODOMETER_KM):
	Unique values: 
(13,)
	Description: 
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
	Value counts: 

		Greatest: 
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64
		Least: 
5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64


Column (PRICE_DOLLAR):
	Unique values: 
(2357,)
	Description: 
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollar, dtype: float64
	Value counts: 

		Greatest: 
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_dollar, dtype: int64
		Least: 
0    1421
1     156
2       3
3       1
5       2
Na

{'odometer_km': 0        150000
 1        150000
 2         70000
 3         70000
 4        150000
 5        150000
 6        150000
 7        150000
 8        150000
 9        150000
 10       150000
 11       150000
 12        50000
 13       150000
 14       150000
 15        80000
 16       150000
 17       150000
 18       150000
 19       150000
 20       150000
 21       150000
 22        10000
 23       150000
 24        30000
 25       150000
 26       125000
 27       150000
 28       150000
 29       150000
           ...  
 49970     60000
 49971    150000
 49972    150000
 49973    150000
 49974    150000
 49975    100000
 49976    150000
 49977    150000
 49978    150000
 49979     70000
 49980    125000
 49981    150000
 49982     90000
 49983    150000
 49984    150000
 49985    150000
 49986    125000
 49987     50000
 49988    150000
 49989    150000
 49990     30000
 49991    150000
 49992    125000
 49993    150000
 49994    150000
 49995    100000
 49996    150000

#### Note:
As can be seen, ***odometer_km*** contains reasonable values with no unlikely outliers. Comparing the age of the car to the odometer readings might highlight more unlikely values, but it would be hard to tell given that some cars can sit idle for decades, while others get driven to failure within a few years.

For the ***price_dollar*** column, <code>Series.describe()</code> points out unlikely candidates. The focus is primarily on cars in the top and bottom 10%-brackets (10% was chosen arbitrarily). Let's perform further analysis:

In [27]:
col_detail(0.9, "price_dollar")



Column (PRICE_DOLLAR):
	Unique values: 
(931,)
	Description: 
count    5.000000e+03
mean     6.545228e+04
std      1.520358e+06
min      1.400000e+04
25%      1.649000e+04
50%      1.990000e+04
75%      2.639250e+04
max      1.000000e+08
Name: price_dollar, dtype: float64
	Value counts: 

		Greatest: 
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_dollar, dtype: int64
		Least: 
14000    51
14050     1
14100     2
14190     2
14199     1
Name: price_dollar, dtype: int64


{'price_dollar': 39770       14000
 10768       14000
 27055       14000
 25018       14000
 6481        14000
 29578       14000
 39693       14000
 7316        14000
 26387       14000
 21125       14000
 39775       14000
 40971       14000
 19164       14000
 35675       14000
 32097       14000
 6035        14000
 37154       14000
 832         14000
 15685       14000
 18666       14000
 2373        14000
 517         14000
 43099       14000
 173         14000
 39393       14000
 17334       14000
 16189       14000
 18502       14000
 34162       14000
 13554       14000
            ...   
 18509      169000
 32840      169999
 11433      175000
 17140      180000
 20351      190000
 28090      194000
 43668      197000
 40918      198000
 37840      220000
 38299      250000
 47337      259000
 12682      265000
 35923      295000
 34723      299000
 14715      345000
 36818      350000
 37585      999990
 514        999999
 43049      999999
 22947     1234566
 7814      1300

#### Note:
The upper bracket has a mean of \$65,452, undoubtedly skewed by the clear outliers of \$900,000+, as seen by the 75%-mark of $26,392. It can be argued that anything above this mean is an outlier, seeing as eBay hardly seems like the appropriate place to buy high-end vehicles, but just by looking at the distribtion of price, the obvious "fake" entries are over \$300,000:

In [28]:
outliers_upper = autos.loc[autos["price_dollar"]>300000]
outliers_upper.shape

(16, 16)

#### Note:
Now for the lower bracket, let's look at the bottom 25%-bracket:

In [29]:
col_detail(0.25, "price_dollar")



Column (PRICE_DOLLAR):
	Unique values: 
(271,)
	Description: 
count    12500.000000
mean       554.769920
std        338.333247
min          0.000000
25%        300.000000
50%        599.000000
75%        850.000000
max       1100.000000
Name: price_dollar, dtype: float64
	Value counts: 

		Greatest: 
1100    338
1099     44
1098      1
1095      3
1090      4
Name: price_dollar, dtype: int64
		Least: 
0    1421
1     156
2       3
3       1
5       2
Name: price_dollar, dtype: int64


{'price_dollar': 18089       0
 43923       0
 15225       0
 29499       0
 8445        0
 38832       0
 43925       0
 8438        0
 35821       0
 31332       0
 12819       0
 33620       0
 15208       0
 2481        0
 4266        0
 27020       0
 35819       0
 31336       0
 29482       0
 47861       0
 21253       0
 37021       0
 29503       0
 21256       0
 2470        0
 24871       0
 31264       0
 1220        0
 24900       0
 21171       0
          ... 
 5215     1100
 29424    1100
 29917    1100
 25385    1100
 46989    1100
 43543    1100
 16591    1100
 16143    1100
 38921    1100
 30224    1100
 41539    1100
 2165     1100
 36602    1100
 12308    1100
 25159    1100
 8998     1100
 5904     1100
 20201    1100
 21334    1100
 2171     1100
 44077    1100
 36785    1100
 49186    1100
 1965     1100
 38800    1100
 25408    1100
 43761    1100
 43894    1100
 11672    1100
 21056    1100
 Name: price_dollar, Length: 12500, dtype: int64}

#### Note:
It is hard to make a cutoff point here, but I will choose all cars less priced at less than \$100:

In [30]:
outliers_lower = autos.loc[autos["price_dollar"]<100]
outliers_lower

date_crawled                                               name  \
25    2016-03-21 21:56:18  ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
27    2016-03-27 18:45:01             hat_einer_ahnung_mit_ford_galaxy_hilfe   
30    2016-03-14 11:47:31                         peugeot_206_unfallfahrzeug   
55    2016-03-07 02:47:54                      mercedes_e320_amg_zu_tauschen   
64    2016-04-05 07:36:19                   autotransportabschlepp_schlepper   
71    2016-03-28 19:39:35  suche_opel_astra_fcorsa_oder_kadett_e_mit_rest...   
80    2016-03-09 15:57:57  nissan_primera_hatchback_1_6_16v_73_kw99psbj03...   
87    2016-03-29 23:37:22                      bmw_520_e39_zum_ausschlachten   
99    2016-04-05 09:48:54                       peugeot_207_cccabrio_bj_2011   
118   2016-03-12 05:03:00  vw_sharan_v6_204_ps_karosse_rohkarosse_mit_pap...   
146   2016-03-22 23:59:28                                    ford_fiesta_rot   
167   2016-04-02 19:43:45  suche_vw_multivan_innenausstattung_set_oder_te...   
180   2016-03-19 10:50:25                                       zu_verkaufen   
226   2016-03-25 23:52:12                        porsche_911_s_targa67er_swb   
234   2016-03-05 16:56:22                                   fiat_punto_sport   
245   2016-03-30 12:45:44                                     renault_twingo   
248   2016-03-08 22:57:53                             vw_passat_zu_verkaufen   
259   2016-04-03 23:49:58                         guenstiges_autoauch_defekt   
301   2016-03-08 20:37:59                         kaufe_alle_autos_bietet_an   
323   2016-03-15 16:54:59                                w220_unfallfahrzeug   
327   2016-04-03 13:49:21                 bmw_cabrio_318ibmw_318_tds_touring   
344   2016-03-22 11:25:18  verkaufe_hier_mein_mein_schoener_honda_crx_in_...   
347   2016-04-02 11:48:02                                    volvo_tp21sugga   
366   2016-03-21 20:36:25         ford_escort_express_75lkw_zulussungdiesel_   
389   2016-03-08 11:38:52                                    vw_polo_classic   
418   2016-03-29 14:43:24             fiat_scudo_8_sitzer_busdiesel_jtd80_kw   
427   2016-03-12 11:44:47                    top_zustand_gruneplakette_euro4   
430   2016-03-18 23:52:40                                  winterraeder_ford   
449   2016-04-03 20:59:59                           tausche_gegen_ein_smart.   
487   2016-03-14 21:41:36  mag_mich_keiner_haben_opel_omega_b_caravan_2.0...   
...                   ...                                                ...   
49082 2016-03-13 17:06:58         passat_35i_b4_in_candy_weiss_original_lack   
49240 2016-03-10 09:54:48    suche_7_sitzer_espace_oder_biete_s.beschreibung   
49254 2016-03-21 10:49:57                           corrado_g60_mit_2e_motor   
49301 2016-03-20 20:52:40                           golf_3_variant_kombi_1.8   
49342 2016-03-09 20:47:01  audi_80_b4_s_color_und_bmw_318ti_compact_tausc...   
49350 2016-03-19 23:57:15                                     golf_ii_diesel   
49377 2016-03-20 00:58:36             tausche_bmw_318i_facelift_model_143_ps   
49404 2016-04-04 03:03:16                                golf_2_bbs_leder_kw   
49408 2016-03-30 20:57:06                                         opel_astra   
49479 2016-04-01 23:55:01                             tausch_mit_frischen_tv   
49496 2016-03-26 13:55:28                                        bmw_e39_520   
49504 2016-03-13 19:42:57                                 meriva_lpg_1_6_16v   
49507 2016-03-16 17:38:32                 bmw_528i_e39_schaltgetriebe_unfall   
49525 2016-03-09 17:59:02              bmw_e36_318_i_limo_ringtool_ps_tuning   
49528 2016-03-28 19:57:41           bmw_e36316i_compact_m_paket_schlachtfest   
49538 2016-03-14 10:55:27                           golf_3_mit_tuevpreiswert   
49551 2016-03-16 23:40:18          golf_iv_an_bastler_zuverkaufenreserviert_   
49614 2016-03-09 17:49:39      verkauf_meineaudia3gegen_tauschen_mit_diesel.   
49728 2016-03

In [31]:
outliers = pd.concat([outliers_lower, outliers_upper])
autos.drop(index=outliers.index, inplace=True)

In [32]:
col_detail(0.0, *[c for c in autos.columns])



Column (DATE_CRAWLED):
	Unique values: 
(46570,)
	Description: 
count                   48222
unique                  46570
top       2016-03-23 18:39:34
freq                        3
first     2016-03-05 14:06:30
last      2016-04-07 14:36:56
Name: date_crawled, dtype: object
	Value counts: 

		Greatest: 
2016-04-07 14:36:56    1
2016-04-07 14:36:55    1
2016-04-07 14:36:44    1
2016-04-07 14:30:26    1
2016-04-07 14:30:09    1
Name: date_crawled, dtype: int64
		Least: 
2016-03-05 14:06:30    1
2016-03-05 14:06:40    1
2016-03-05 14:07:04    1
2016-03-05 14:07:08    1
2016-03-05 14:07:21    1
Name: date_crawled, dtype: int64


Column (NAME):
	Unique values: 
(36242,)
	Description: 
count           48222
unique          36242
top       ford_fiesta
freq               93
Name: name, dtype: object
	Value counts: 

		Greatest: 
zwischenloesung                                   1
zweimassenschwungradraederzws_crafter_sprinter    1
zwei_fahrzeuges_gegen_eine                        1
zwar_a

{'abtest': 0        control
 1        control
 2           test
 3        control
 4           test
 5           test
 6           test
 7        control
 8           test
 9        control
 10          test
 11       control
 12       control
 13       control
 14          test
 15          test
 16          test
 17          test
 18       control
 19       control
 20          test
 21          test
 22       control
 23       control
 24       control
 26       control
 28       control
 29          test
 31          test
 32       control
           ...   
 49968       test
 49969    control
 49970    control
 49971       test
 49972    control
 49973    control
 49975    control
 49976       test
 49977    control
 49978    control
 49979       test
 49980    control
 49981    control
 49982    control
 49983       test
 49985    control
 49986    control
 49987    control
 49988    control
 49989       test
 49990       test
 49991    control
 49992    control
 49993    control


#### Registration:
Next for the ***date_registration*** column, which contains impossible values. Any car registered after its listing was seen is inaccurate; all cars registered after 2016. On the other end of the spectrum, the first official license plate was created and given out in 1903, thus that will be the lower bound. Let's analyze these outliers:

In [33]:
outliers_lower = autos[autos["date_registration"]<dt(1903, 1, 1)]
outliers_upper = autos[autos["date_registration"]>dt(2016, 12, 31)]
outliers = pd.concat([outliers_lower, outliers_upper])

display(outliers_lower)
display(outliers_upper)

date_crawled        name  price_dollar   abtest type gearbox  \
10556 2016-04-01 06:02:10  unfal_auto           450  control  NaN     NaN   
32585 2016-04-02 16:56:39  unfal_auto           450  control  NaN     NaN   

       power_ps model  odometer_km fuel       brand unrepaired_damage  \
10556      1800   NaN         5000  NaN  mitsubishi                no   
32585      1800   NaN         5000  NaN  mitsubishi                no   

      date_ad_created  postal_code          last_update date_registration  
10556      2016-04-01        63322  2016-04-01 09:42:30        1800-02-01  
32585      2016-04-02        63322  2016-04-04 14:46:21        1800-02-01

date_crawled                                               name  \
10    2016-03-15 01:41:36                       vw_golf_tuning_in_siber/grau   
65    2016-04-04 19:30:39                      ford_fiesta_zum_ausschlachten   
113   2016-04-03 14:58:29                              golf_4_anfaenger_auto   
164   2016-03-13 20:39:16  opel_merivanur_76000_kmunfallfreischeckheftgep...   
197   2016-04-05 10:36:24                              vw_polo_9n_an_bastler   
253   2016-03-27 13:25:18              ford_mondeo_gas_anlage_mit_tv_04.2017   
348   2016-03-17 20:58:24  vw_beetle_1.8turbo_mit_vollausstattung_und_seh...   
390   2016-03-25 12:59:06             fiat_bertone_x_1_9x_1/9x19x_19x1_9x_19   
457   2016-03-15 17:48:39  mercedes_benz_w203_c200_kombi_kompressor_c_kla...   
477   2016-03-31 08:53:34                                          citron_c4   
545   2016-03-22 18:50:53                                          bmw_x5e53   
548   2016-03-10 17:44:57            peugeot_206_1.9_hdi_4_tueren_tuev_02.17   
557   2016-03-16 03:36:34                           smart_fortwo_passion_mhd   
568   2016-03-18 09:48:44              prima_anfngerauto_smart_fortwo_cabrio   
577   2016-04-03 14:49:09                                  volkswagen_golf_3   
582   2016-03-31 11:57:46                                    v10_tdi_touareg   
654   2016-04-01 15:48:34              ich_verkaufe_meinen_gepflegten_daewoo   
740   2016-04-02 19:49:46                             seat_leon_zu_verkaufen   
754   2016-03-26 13:47:56                         ford_ka_kleinwagen_bj_2008   
758   2016-03-12 18:42:33                                honda_civic_mit_tv_   
765   2016-03-26 11:52:53                    bmw_3er_limousine_mit_1_jahr_tv   
871   2016-04-04 19:38:38                                   citroen_c2_60_ps   
889   2016-03-12 16:52:19                          verkaufe_opel_corsa_c_1_2   
910   2016-03-15 17:52:50                         renault_kangoo_bj.2007_lpg   
918   2016-03-15 15:53:56                              bmw_114i_sondermodell   
1030  2016-03-24 20:47:08                          ford_galaxydefekt_bastler   
1040  2016-03-30 19:53:16                                         audi_80_b4   
1066  2016-03-06 15:49:03                ford_escort_kombi/turnier_tv_7.2017   
1121  2016-03-26 21:38:21  opel_zafira_a_1.8_z18xe_hu_02/18_164tkm_alu_klima   
1128  2016-03-08 10:55:10                                 audi_a6_2.0_s_line   
...                   ...                                                ...   
48937 2016-03-27 13:54:58                  mazda_3_60tkm_sitzheizung5_tueren   
48963 2016-03-20 23:52:05                  golf_1.4_editiontv_8/17158.000tkm   
48975 2016-03-21 13:44:47                                 renault_espace_iii   
48980 2016-03-28 19:50:38                                       toyota_yaris   
49003 2016-03-12 12:52:08                                 renault_1.hand_2.0   
49023 2016-03-23 00:49:23                                       auto_audi_a3   
49053 2016-03-15 15:57:37                     volkswagen_golf_1.9_tdi_110_ps   
49054 2016-03-11 23:37:19                              skoda_fabia_aus_1hand   
49075 2016-03-18 12:47:25            vw_golf_2_gt_tv_neu_gti_g60_16v_special   
49076 2016-03-25 21:39:21                             omega_b_2.0_16v_136_ps   
49108 2016-03-14 22:37:20                    tdi_20_l_diesel140_psautomatick   
49130 2016-03-09 18:46:05                                      renault_modus   
49154 2016-03-24 17:51:22                     auto_mit_tv_bis_september_2017   
49178 2016-03-24 19:58:43          mazda_323_f_tvneu_2te_hand_viele_neuteile   
49185 2016-03-28 11:45:31              alfa_romeo_156_sportwagon_1_8_t_spark   
49218 2016-04-01 19:48:06          ford_fiesta_1.3_4/5_tuerer_16_monate_tuev   
49245 2016-03-26 08:32:30                         audi_a3_1.8t_tausch_moegl.   
49261 2016-03-10 12:51:40                                        toyota_aygo   
49266 2016-03

In [34]:
autos.drop(index=outliers.index, inplace=True)
autos.shape

(46802, 16)

#### Exploring top 20 most posted brands

In [75]:
brands = autos["brand"].value_counts()
brands[:20]

volkswagen        9915
bmw               5127
opel              5037
mercedes_benz     4527
audi              4050
ford              3258
renault           2215
peugeot           1395
fiat              1206
seat               860
skoda              761
mazda              717
nissan             712
smart              665
citroen            656
toyota             599
hyundai            466
sonstige_autos     449
volvo              425
mini               411
Name: brand, dtype: int64

In [80]:
top20_price = {}
top20_km = {}

for brand in brands.index[:20]:
    top20_price[brand] = autos.loc[autos["brand"]==brand, "price_dollar"].mean()
    top20_km[brand] = autos.loc[autos["brand"]==brand, "odometer_km"].mean()

display(top20_price)
display(top20_km)

{'audi': 9337.918518518518,
 'bmw': 8363.22371757363,
 'citroen': 3800.7118902439024,
 'fiat': 2822.9228855721394,
 'ford': 3768.815224063843,
 'hyundai': 5400.727467811159,
 'mazda': 4104.839609483961,
 'mercedes_benz': 8625.957366909654,
 'mini': 10611.766423357663,
 'nissan': 4751.664325842697,
 'opel': 2989.7621600158823,
 'peugeot': 3103.048745519713,
 'renault': 2479.6988713318283,
 'seat': 4392.048837209302,
 'skoda': 6409.609724047306,
 'smart': 3579.2962406015035,
 'sonstige_autos': 11882.913140311804,
 'toyota': 5151.058430717863,
 'volkswagen': 5411.077155824509,
 'volvo': 4977.24}

{'audi': 129298.76543209876,
 'bmw': 132722.83986736884,
 'citroen': 119474.08536585367,
 'fiat': 117089.55223880598,
 'ford': 124370.77961939841,
 'hyundai': 107070.81545064377,
 'mazda': 124672.24546722455,
 'mercedes_benz': 130970.84161696488,
 'mini': 88637.46958637469,
 'nissan': 118370.78651685393,
 'opel': 129358.7452848918,
 'peugeot': 127114.6953405018,
 'renault': 128081.26410835214,
 'seat': 121732.55813953489,
 'skoda': 110906.70170827858,
 'smart': 100135.33834586466,
 'sonstige_autos': 90222.7171492205,
 'toyota': 116168.61435726211,
 'volkswagen': 128877.96268280383,
 'volvo': 138635.29411764705}

In [87]:
price_series = pd.Series(top20_price)
km_series = pd.Series(top20_km)

top20 = pd.DataFrame({"mean_price": price_series, "mean_odometer_km": km_series})
top20

mean_odometer_km    mean_price
audi               129298.765432   9337.918519
bmw                132722.839867   8363.223718
citroen            119474.085366   3800.711890
fiat               117089.552239   2822.922886
ford               124370.779619   3768.815224
hyundai            107070.815451   5400.727468
mazda              124672.245467   4104.839609
mercedes_benz      130970.841617   8625.957367
mini                88637.469586  10611.766423
nissan             118370.786517   4751.664326
opel               129358.745285   2989.762160
peugeot            127114.695341   3103.048746
renault            128081.264108   2479.698871
seat               121732.558140   4392.048837
skoda              110906.701708   6409.609724
smart              100135.338346   3579.296241
sonstige_autos      90222.717149  11882.913140
toyota             116168.614357   5151.058431
volkswagen         128877.962683   5411.077156
volvo              138635.294118   4977.240000

#### Note:
Some possible inferences: (high odometer, high price) -> brand depreciates slowly, (low odometer, low price) -> 